src/parallel.h

```diff
@@ -168,132 +170,255 @@ struct CopyIfScanBody {
-  bool inTmp = false;
-
-  SortedRange(T *input, T *tmp, SizeType offset = 0, SizeType length = 0)
-      : input(input), tmp(tmp), offset(offset), length(length) {}
-  SortedRange(SortedRange<T, SizeType> &r, tbb::split)
-      : input(r.input), tmp(r.tmp) {}
-  // FIXME: no idea why thread sanitizer reports data race here
-#if defined(__has_feature)
-#if __has_feature(thread_sanitizer)
-  __attribute__((no_sanitize("thread")))
-#endif
-#endif
-  void
-  operator()(const tbb::blocked_range<SizeType> &range) {
-    SortedRange<T, SizeType> rhs(input, tmp, range.begin(),
-                                 range.end() - range.begin());
-    rhs.inTmp =
-        LSB_radix_sort(input + rhs.offset, tmp + rhs.offset, rhs.length);
-    if (length == 0)
-      *this = rhs;
-    else
-      join(rhs);
+template <typename KeyFn>
+struct ByteGetter {
+  KeyFn &keyfn;
+  ByteGetter(KeyFn &keyfn) : keyfn(keyfn) {}
+  template <typename I>
+  size_t operator()(I curr, int i) const {
+    return static_cast<size_t>((keyfn(*curr) >> (8 * i)) & 0xFF);
   }
-  bool swapBuffer() const {
-    T *src = input, *target = tmp;
-    if (inTmp) std::swap(src, target);
-    copy(src + offset, src + offset + length, target + offset);
-    return !inTmp;
+};
+
+template <typename U>
+struct RadixBlockMetadata {
+  U localMin;
+  U localMax;
+  U prefixMax;
+  U suffixMin;
+  bool sorted;
+  bool canSkip;
+};
+
+template <typename I, typename J, typename KeyFn>
+void radix_sort_with_key_rec(I start, I end, J dest, int bytes, KeyFn &keyfn,
+                             bool writeback) {
+  using T = std::remove_reference_t<decltype(*start)>;
+  using U = std::invoke_result_t<KeyFn, T>;
+  constexpr size_t MIN_BLOCK_SIZE = 10'000;
+  size_t n = std::distance(start, end);
+
+  if ((n <= kSeqThreshold && bytes >= 1) || (n < 2'000)) {
+    if (writeback) {
+      std::stable_sort(start, end, [&keyfn](const T &a, const T &b) {
+        return keyfn(a) < keyfn(b);
+      });
+    } else {
+      std::copy(start, end, dest);
+      std::stable_sort(dest, dest + n, [&keyfn](const T &a, const T &b) {
+        return keyfn(a) < keyfn(b);
+      });
+    }
+    return;
   }
-  void join(const SortedRange<T, SizeType> &rhs) {
-    if (inTmp != rhs.inTmp) {
-      if (length < rhs.length)
-        inTmp = swapBuffer();
+  auto getByte = ByteGetter(keyfn);
+
+  unsigned int hardware_concurrency = std::thread::hardware_concurrency();
+  size_t blocks =
+      std::max(std::min(hardware_concurrency,
+                        static_cast<unsigned int>(n / MIN_BLOCK_SIZE)),
+               1u);
+#if MANIFOLD_PAR == 1
+  if (blocks > 1) {
+    size_t blockSize = n / blocks;
+    std::vector<std::array<size_t, 256>> hists(blocks + 1);
+
+    // local min, local max, prefix max, suffix min, sorted, can skip
+    std::vector<RadixBlockMetadata<U>> metadata(blocks);
+
+    auto taskgroup = tbb::task_group();
+    for (size_t i = 0; i < blocks; ++i) {
+      taskgroup.run([&, i]() {
```

Comment:
We compute per-block metadata on the fly, which can be used to determine if the prefix or suffix of the input is already sorted and do not require further processing.